sql/lease.go

```diff
@@ -0,0 +1,511 @@
+		// invariant that no new leases for desc.Version-1 will be granted once
+		// desc.Version exists.
+		expectedVersion := desc.Version
+		err = s.db.Txn(func(txn *client.Txn) error {
+			// Re-read the current version of the table descriptor, this time
+			// transactionally.
+			if err := txn.GetProto(descKey, desc); err != nil {
+				return err
+			}
+			if expectedVersion != desc.Version {
+				// The version changed out from under us. Someone else must be
+				// performing a schema change operation.
+				if log.V(3) {
+					log.Infof("publish (version changed): %d != %d", expectedVersion, desc.Version)
+				}
+				return errLeaseVersionChanged
+			}
+
+			// Run the update closure which is intended to perform a single step in a
+			// multi-step schema change operation.
+			if err := update(desc); err != nil {
+				return err
+			}
+
+			// Bump the version and modification time.
+			desc.Version = desc.Version + 1
+			desc.ModificationTime = now
+			if log.V(3) {
+				log.Infof("publish: descID=%d version=%d mtime=%s", desc.ID, desc.Version, now.GoTime())
+			}
+
+			// Write the updated descriptor.
+			b := &client.Batch{}
+			b.Put(descKey, desc)
+			txn.SetSystemDBTrigger()
+			return txn.CommitInBatch(b)
+		})
+
+		if err != errLeaseVersionChanged {
+			return err
+		}
+	}
+
+	panic("not reached")
+}
+
+// countLeases returns the number of unexpired leases for a particular version
+// of a descriptor.
+func (s LeaseStore) countLeases(descID ID, version uint32, expiration int64) (int, error) {
+	var count int
+	err := s.db.Txn(func(txn *client.Txn) error {
+		p := planner{txn: txn, user: security.RootUser}
+
+		const countLeases = `SELECT COUNT(version) FROM system.lease ` +
+			`WHERE descID = %d AND version = %d AND expiration > '%s'::timestamp`
+		sql := fmt.Sprintf(countLeases, descID, version, nanosToDTimestamp(expiration))
+		values, err := p.queryRow(sql)
+		if err != nil {
+			return err
+		}
+		count = (int)(values[0].(parser.DInt))
+		return nil
+	})
+	return count, err
+}
+
+// leaseSet maintains an ordered set of LeaseState objects. It supports
+// addition and removal of elements, finding a specific lease, finding the
+// newest lease for a particular version and finding the newest lease for the
+// most recent version.
+type leaseSet struct {
+	// The lease state data is stored in a sorted slice ordered by <version,
+	// expiration>. Ordering is maintained by insert and remove.
+	data []*LeaseState
+}
+
+func (l *leaseSet) String() string {
+	var buf bytes.Buffer
+	for i, s := range l.data {
+		if i > 0 {
+			buf.WriteString(" ")
+		}
+		buf.WriteString(s.String())
+	}
+	return buf.String()
+}
+
+func (l *leaseSet) insert(s *LeaseState) {
+	i, match := l.findIndex(s.Version, s.expiration)
+	if match {
+		panic("unable to insert duplicate lease")
+	}
+	if i == len(l.data) {
+		l.data = append(l.data, s)
+		return
+	}
+	l.data = append(l.data, nil)
```

Comment:
Ah, right. I guess we usually nest `append`s in this case, but this is likely more performant. Either way.